sql-plugin/src/main/330+/scala/org/apache/spark/sql/rapids/shims/intervalExpressions.scala

```diff
@@ -184,6 +186,113 @@ object IntervalUtils extends Arm {
+        withResource(Scalar.fromInt(0)) { zero =>
+          withResource(cv.equalTo(zero)) { isZero =>
+            BoolUtils.isAnyValidTrue(isZero)
+          }
+        }
+    }
+  }
+
+  /**
+   * Divide p by q rounding with the HALF_UP mode.
+   * Logic is round(p.asDecimal / q, HALF_UP).
+   *
+   * It's equivalent to
+   * com.google.common.math.LongMath.divide(p, q, RoundingMode.HALF_UP):
+   * long div = p / q; // throws if q == 0
+   * long rem = p - q * div; // equals p % q
+   * // signum is 1 if p and q are both non-negative or both negative, and -1 otherwise.
+   * int signum = 1 | (int) ((p xor q) >> (Long.SIZE - 1));
+   * long absRem = abs(rem);
+   * long cmpRemToHalfDivisor = absRem - (abs(q) - absRem);
+   * increment = cmpRemToHalfDivisor >= 0
+   * return increment ? div + signum : div;
+   *
+   * @param p integer(int/long) cv or scalar
+   * @param q integer(byte/short/int/long) cv or Scalar, if p is scala, q will not be scala
+   * @return decimal 128 cv of p / q
+   */
+  def divWithHalfUpModeWithOverflowCheck(p: BinaryOperable, q: BinaryOperable): ColumnVector = {
+    // 1. overflow check q is 0
+    if (IntervalUtils.hasZero(q)) {
+      throw new ArithmeticException("overflow: interval / zero")
+    }
+
+    // 2. overflow check (p == min(int min or long min) && q == -1)
+    val min = p.getType match {
+      case DType.INT32 => Int.MinValue.toLong
+      case DType.INT64 => Long.MinValue
+      case _ => throw new IllegalStateException()
+    }
+    withResource(Scalar.fromLong(min)) { minScalar =>
+      withResource(Scalar.fromLong(-1L)) { negOneScalar =>
+        (p, q) match {
+          case (lCv: ColumnVector, rCv: ColumnVector) =>
+            withResource(lCv.equalTo(minScalar)) { isMin =>
+              withResource(rCv.equalTo(negOneScalar)) { isNegOne =>
+                withResource(isMin.and(isNegOne)) { invalid =>
+                  if (BoolUtils.isAnyValidTrue(invalid)) {
+                    throw new ArithmeticException("overflow occurs")
+                  }
+                }
+              }
+            }
+          case (lCv: ColumnVector, rS: Scalar) =>
+            withResource(lCv.equalTo(minScalar)) { isMin =>
+              if (getLong(rS) == -1L && BoolUtils.isAnyValidTrue(isMin)) {
+                throw new ArithmeticException("overflow occurs")
+              }
+            }
+          case (lS: Scalar, rCv: ColumnVector) =>
+            withResource(rCv.equalTo(negOneScalar)) { isNegOne =>
+              if (getLong(lS) == min && BoolUtils.isAnyValidTrue(isNegOne)) {
+                throw new ArithmeticException("overflow occurs")
+              }
+            }
+          case (lS: Scalar, rS: Scalar) =>
+            getLong(lS) == min && getLong(rS) == -1L
+        }
+      }
+    }
+
+    // 3. round(p.asDecimal / q)
+    val dT = DType.create(DType.DTypeEnum.DECIMAL128, -1)
+    val leftDecimal = p match {
+      case pCv: ColumnVector => pCv.castTo(dT)
```

Comment:
Is pre-casting really necessary, or will this happen naturally when we specify the desired output type in the `.div` call below?  If cudf is going to pre-cast for us during the div operation, it would be nice to avoid manifesting the intermediate result explicitly.

Reply:
Cudf does not do this for us. If so, it will improve the performance.

```
    val dT = DType.create(DType.DTypeEnum.DECIMAL128, -1)
    val cv1 = ColumnVector.fromLongs(5L)
    val cv2 = ColumnVector.fromLongs(3L)
    println(cv1.div(cv2, dT).getScalarElement(0))
```
get Scalar{type=DECIMAL128 scale:-1 **value=1.0**}

But for the `long_cv.mul(long_cv, DECIMAL128(scale = 0))`, we can do it, the only difference is the `scale = 0` instead of `scale = 
 1`

I think we can file a `feature request`  to cuDF, what do you think?

Reply:
Yeah I think cudf will only upcast to the "natural" type of the two operands and the operator then cast at the end to the desired type.  Since C++ won't naturally upcast to __int128_t for this operator, I don't think it's reasonable to ask cudf to add this feature.  I'm sure they would just push back and say we need to upcast if we need that behavior.  Otherwise it would add quite a bit of complexity to all their binary operators for arguably a fairly specialized type of behavior.